File: core/player.py
import time
import threading
from typing import List, Callable, Optional, Tuple
from enum import Enum
from .actions import Action, ActionType


class PlayerState(Enum):
    IDLE = "idle"
    PLAYING = "playing"
    PAUSED = "paused"
    STOPPED = "stopped"


class WindowOffsetProvider:
    def __init__(self, hwnd: int = 0, window_utils=None):
        self._hwnd = hwnd
        self._window_utils = window_utils
        self._last_offset: Optional[Tuple[int, int]] = None
        self._last_error: Optional[str] = None
    
    def set_hwnd(self, hwnd: int):
        self._hwnd = hwnd
    
    def set_window_utils(self, window_utils):
        self._window_utils = window_utils
    
    def get_current_offset(self) -> Tuple[Optional[Tuple[int, int]], Optional[str]]:
        if not self._hwnd:
            return None, None
        
        if not self._window_utils:
            return None, "窗口工具不可用"
        
        try:
            window_info = self._window_utils.get_window_by_hwnd(self._hwnd)
            if not window_info:
                return None, f"窗口不存在或已关闭 (句柄: {self._hwnd})"
            
            self._last_offset = (window_info.x, window_info.y)
            self._last_error = None
            return self._last_offset, None
        except Exception as e:
            self._last_error = str(e)
            return None, self._last_error
    
    def validate_window(self) -> Tuple[bool, str]:
        if not self._hwnd:
            return True, ""
        
        if not self._window_utils:
            return True, ""
        
        try:
            window_info = self._window_utils.get_window_by_hwnd(self._hwnd)
            if not window_info:
                return False, f"窗口不存在或已关闭 (句柄: {self._hwnd})"
            return True, ""
        except Exception as e:
            return False, str(e)


class Player:
    def __init__(self, tab_key: str = "", local_group_manager=None):
        self._tab_key = tab_key
        self._state = PlayerState.IDLE
        self._state_lock = threading.Lock()
        self.actions: List[Action] = []
        self.current_index: int = 0
        self.speed: float = 1.0
        self.repeat_count: int = 1
        self.current_repeat: int = 0
        self.infinite_loop: bool = False
        self.timeout_seconds: float = 0
        self._local_group_manager = local_group_manager
        
        self._thread: Optional[threading.Thread] = None
        self._pause_event = threading.Event()
        self._stop_flag = False
        self._window_offset: Optional[Tuple[int, int]] = None
        self._window_title: str = ""
        self._start_time: float = 0
        
        self._window_offset_provider: Optional[WindowOffsetProvider] = None
        self._window_hwnd: int = 0
        self._window_utils = None
        
        self._callbacks = {
            'on_action_start': [],
            'on_action_end': [],
            'on_state_changed': [],
            'on_progress': [],
            'on_error': [],
            'on_finished': [],
            'on_repeat_changed': [],
            'on_window_error': []
        }
    
    def set_local_group_manager(self, manager):
        self._local_group_manager = manager
    
    def get_local_group_manager(self):
        return self._local_group_manager
    
    @property
    def state(self) -> PlayerState:
        with self._state_lock:
            return self._state
    
    @state.setter
    def state(self, value: PlayerState):
        with self._state_lock:
            self._state = value
    
    def add_callback(self, event: str, callback: Callable):
        if event in self._callbacks:
            self._callbacks[event].append(callback)
    
    def remove_callback(self, event: str, callback: Callable):
        if event in self._callbacks and callback in self._callbacks[event]:
            self._callbacks[event].remove(callback)
    
    def _emit(self, event: str, *args, **kwargs):
        for callback in self._callbacks.get(event, []):
            try:
                callback(*args, **kwargs)
            except Exception as e:
                print(f"Callback error: {e}")
    
    def set_actions(self, actions: List[Action]):
        if self.state != PlayerState.IDLE:
            return
        self.actions = actions.copy()
    
    def set_speed(self, speed: float):
        self.speed = max(0.1, min(10.0, speed))
    
    def set_repeat_count(self, count: int):
        self.repeat_count = max(1, count)
    
    def set_infinite_loop(self, enabled: bool):
        self.infinite_loop = enabled
    
    def set_timeout(self, seconds: float):
        self.timeout_seconds = max(0, seconds)
    
    def set_window_offset(self, offset: Optional[Tuple[int, int]]):
        self._window_offset = offset
    
    def set_window_title(self, title: str):
        self._window_title = title
    
    def set_window_hwnd(self, hwnd: int, window_utils=None):
        self._window_hwnd = hwnd
        self._window_utils = window_utils
        
        if hwnd and window_utils:
            if not self._window_offset_provider:
                self._window_offset_provider = WindowOffsetProvider(hwnd, window_utils)
            else:
                self._window_offset_provider.set_hwnd(hwnd)
                self._window_offset_provider.set_window_utils(window_utils)
        else:
            self._window_offset_provider = None
    
    def _get_realtime_window_offset(self) -> Tuple[Optional[Tuple[int, int]], Optional[str]]:
        if not self._window_offset_provider:
            return self._window_offset, None
        
        offset, error = self._window_offset_provider.get_current_offset()
        if error:
            return None, error
        
        return offset, None
    
    def _validate_window_before_action(self, action: Action) -> Tuple[bool, Optional[str]]:
        if not self._window_offset_provider:
            return True, None
        
        needs_window = action.action_type in [
            ActionType.MOUSE_CLICK_RELATIVE,
            ActionType.MOUSE_MOVE_RELATIVE,
            ActionType.MOUSE_DRAG,
        ]
        
        if action.use_relative_coords:
            needs_window = True
        
        if not needs_window:
            return True, None
        
        is_valid, error = self._window_offset_provider.validate_window()
        if not is_valid:
            return False, error
        
        return True, None
    
    def _activate_window_before_action(self, action: Action):
        if not self._window_hwnd or not self._window_utils:
            return
        
        needs_window = action.action_type in [
            ActionType.MOUSE_CLICK_RELATIVE,
            ActionType.MOUSE_MOVE_RELATIVE,
            ActionType.MOUSE_DRAG,
        ]
        
        if action.use_relative_coords:
            needs_window = True
        
        if not needs_window:
            return
        
        try:
            self._window_utils.activate_window(self._window_hwnd)
            time.sleep(0.05)
        except Exception as e:
            print(f"[激活窗口失败] {e}")
    
    def play(self):
        if self.state == PlayerState.PLAYING:
            return
        
        if not self.actions:
            return
        
        if self.state == PlayerState.PAUSED:
            self.state = PlayerState.PLAYING
            self._pause_event.set()
            self._emit('on_state_changed', self.state)
            return
        
        self._ensure_target_window_exists()
        
        self.state = PlayerState.PLAYING
        self.current_index = 0
        self.current_repeat = 0
        self._stop_flag = False
        self._pause_event.set()
        self._start_time = time.time()
        
        self._thread = threading.Thread(target=self._run, daemon=True)
        self._thread.start()
        
        self._emit('on_state_changed', self.state)
    
    def _ensure_target_window_exists(self):
        if not self._window_title:
            return
        
        try:
            import win32gui
            
            from core.command_manager import CommandManager
            cmd_manager = CommandManager.get_instance()
            
            commands = cmd_manager.get_all_commands()
            
            for cmd in commands:
                if cmd.window_title_pattern and cmd.window_title_pattern.lower() in self._window_title.lower():
                    windows_found = []
                    
                    def enum_callback(hwnd, _):
                        if win32gui.IsWindowVisible(hwnd):
                            title = win32gui.GetWindowText(hwnd)
                            if cmd.window_title_pattern.lower() in title.lower():
                                windows_found.append(hwnd)
                        return True
                    
                    win32gui.EnumWindows(enum_callback, None)
                    
                    if not windows_found:
                        success, message, already_running = cmd_manager.check_and_launch(cmd.id)
                        if success and not already_running:
                            time.sleep(2)
                    break
        except Exception:
            pass
    
    def pause(self):
        if self.state == PlayerState.PLAYING:
            self.state = PlayerState.PAUSED
            self._pause_event.clear()
            self._emit('on_state_changed', self.state)
    
    def resume(self):
        if self.state == PlayerState.PAUSED:
            self.state = PlayerState.PLAYING
            self._pause_event.set()
            self._emit('on_state_changed', self.state)
    
    def toggle_pause(self) -> PlayerState:
        with self._state_lock:
            if self._state == PlayerState.PLAYING:
                self._state = PlayerState.PAUSED
                self._pause_event.clear()
                result_state = self._state
            elif self._state == PlayerState.PAUSED:
                self._state = PlayerState.PLAYING
                self._pause_event.set()
                result_state = self._state
            else:
                result_state = self._state
        
        self._emit('on_state_changed', result_state)
        return result_state
    
    def stop(self):
        current_state = self.state
        if current_state in [PlayerState.PLAYING, PlayerState.PAUSED]:
            self._stop_flag = True
            self._pause_event.set()
            self.state = PlayerState.STOPPED
            self._emit('on_state_changed', self.state)
    
    def stop_and_wait(self, timeout: float = 2.0) -> bool:
        current_state = self.state
        if current_state in [PlayerState.PLAYING, PlayerState.PAUSED]:
            self._stop_flag = True
            self._pause_event.set()
            
            if self._thread and self._thread.is_alive():
                self._thread.join(timeout=timeout)
                if self._thread.is_alive():
                    return False
            
            self.state = PlayerState.IDLE
        return True
    
    def _interruptible_sleep(self, seconds: float):
        end_time = time.time() + seconds
        while time.time() < end_time:
            if self._stop_flag:
                return
            sleep_time = min(0.05, end_time - time.time())
            if sleep_time > 0:
                time.sleep(sleep_time)
    
    @property
    def tab_key(self) -> str:
        return self._tab_key
    
    def _run(self):
        completed_actions = 0
        repeat_count = 0
        
        while True:
            if self._stop_flag:
                self.state = PlayerState.IDLE
                self._emit('on_state_changed', self.state)
                self._emit('on_finished', False)
                return
            
            if self.timeout_seconds > 0:
                elapsed = time.time() - self._start_time
                if elapsed >= self.timeout_seconds:
                    self.state = PlayerState.IDLE
                    self._emit('on_state_changed', self.state)
                    self._emit('on_finished', False)
                    return
            
            if not self.infinite_loop and repeat_count >= self.repeat_count:
                break
            
            self.current_repeat = repeat_count
            self._emit('on_repeat_changed', repeat_count + 1)
            
            for i, action in enumerate(self.actions):
                if self._stop_flag:
                    self.state = PlayerState.IDLE
                    self._emit('on_state_changed', self.state)
                    self._emit('on_finished', False)
                    return
                
                if self.timeout_seconds > 0:
                    elapsed = time.time() - self._start_time
                    if elapsed >= self.timeout_seconds:
                        self.state = PlayerState.IDLE
                        self._emit('on_state_changed', self.state)
                        self._emit('on_finished', False)
                        return
                
                self._pause_event.wait()
                
                if self._stop_flag:
                    self.state = PlayerState.IDLE
                    self._emit('on_state_changed', self.state)
                    self._emit('on_finished', False)
                    return
                
                self.current_index = i
                
                if not action.check_condition():
                    print(f"[条件跳过] {action.description} - 条件不满足: {action.condition}")
                    completed_actions += 1
                    self._emit('on_progress', -1, i, repeat_count)
                    continue
                
                is_valid, window_error = self._validate_window_before_action(action)
                if not is_valid:
                    self._emit('on_window_error', action, i, window_error)
                    self.state = PlayerState.IDLE
                    self._emit('on_state_changed', self.state)
                    self._emit('on_finished', False)
                    return
                
                current_offset, offset_error = self._get_realtime_window_offset()
                if offset_error:
                    self._emit('on_window_error', action, i, offset_error)
                    self.state = PlayerState.IDLE
                    self._emit('on_state_changed', self.state)
                    self._emit('on_finished', False)
                    return
                
                if current_offset is None:
                    current_offset = self._window_offset
                
                self._activate_window_before_action(action)
                
                adjusted_delay_before = action.delay_before / self.speed if self.speed > 0 else action.delay_before
                adjusted_delay_after = action.delay_after / self.speed if self.speed > 0 else action.delay_after
                
                if adjusted_delay_before > 0:
                    self._interruptible_sleep(adjusted_delay_before)
                
                if self._stop_flag:
                    self.state = PlayerState.IDLE
                    self._emit('on_state_changed', self.state)
                    self._emit('on_finished', False)
                    return
                
                self._emit('on_action_start', action, i)
                
                try:
                    success = action.execute(window_offset=current_offset, should_stop=lambda: self._stop_flag, local_group_manager=self._local_group_manager)
                    self._emit('on_action_end', action, i, success)
                except Exception as e:
                    self._emit('on_error', action, i, str(e))
                    self._emit('on_action_end', action, i, False)
                
                completed_actions += 1
                self._emit('on_progress', -1, i, repeat_count)
                
                if adjusted_delay_after > 0:
                    self._interruptible_sleep(adjusted_delay_after)
            
            repeat_count += 1
        
        self.state = PlayerState.IDLE
        self._emit('on_state_changed', self.state)
        self._emit('on_finished', True)
    
    def get_state(self) -> PlayerState:
        return self.state
    
    def get_progress(self) -> Tuple[int, int, int]:
        return self.current_index, len(self.actions), self.current_repeat
    
    def is_playing(self) -> bool:
        return self.state == PlayerState.PLAYING
    
    def is_paused(self) -> bool:
        return self.state == PlayerState.PAUSED
    
    def execute_single_action(self, index: int, window_offset: Optional[Tuple[int, int]] = None) -> bool:
        if index < 0 or index >= len(self.actions):
            return False
        
        action = self.actions[index]
        
        is_valid, window_error = self._validate_window_before_action(action)
        if not is_valid:
            self._emit('on_window_error', action, index, window_error)
            return False
        
        current_offset, offset_error = self._get_realtime_window_offset()
        if offset_error:
            self._emit('on_window_error', action, index, offset_error)
            return False
        
        if current_offset is None:
            current_offset = window_offset or self._window_offset
        
        self._activate_window_before_action(action)
        
        adjusted_delay_before = action.delay_before / self.speed if self.speed > 0 else action.delay_before
        adjusted_delay_after = action.delay_after / self.speed if self.speed > 0 else action.delay_after
        
        if adjusted_delay_before > 0:
            time.sleep(adjusted_delay_before)
        
        self._emit('on_action_start', action, index)
        
        try:
            success = action.execute(window_offset=current_offset, should_stop=lambda: self._stop_flag)
            self._emit('on_action_end', action, index, success)
            
            if adjusted_delay_after > 0:
                time.sleep(adjusted_delay_after)
            
            return success
        except Exception as e:
            self._emit('on_error', action, index, str(e))
            self._emit('on_action_end', action, index, False)
            raise

File: crates/engine/src/player.rs
use std::collections::HashMap;
use std::sync::{Arc, atomic::{AtomicBool, Ordering}, Mutex};
use std::thread;
use std::time::{Duration, Instant};

use simplerpa_core::actions::{Action, ActionType};
use simplerpa_core::action_group::LocalActionGroupManager;
use simplerpa_core::config::Config;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum PlayerState {
    Idle,
    Playing,
    Paused,
    Stopped,
}

pub struct Player {
    state: Arc<Mutex<PlayerState>>,
    actions: Vec<Action>,
    pub current_index: usize,
    pub current_repeat: usize,
    pub speed: f64,
    pub repeat_count: i32,
    pub infinite_loop: bool,
    pub timeout_seconds: f64,

    window_offset: Option<(i32, i32)>,
    window_title: String,
    window_hwnd: i64,
    window_run_mode: String,
    local_group_manager: Option<Arc<Mutex<LocalActionGroupManager>>>,

    stop_flag: Arc<AtomicBool>,
    pause_event: Arc<(std::sync::Condvar, std::sync::Mutex<bool>)>,
    start_time: Instant,
    thread_handle: Option<thread::JoinHandle<()>>,

    callbacks: HashMap<String, Vec<Box<dyn Fn(&str) + Send + Sync>>>,
}

impl Player {
    pub fn new() -> Self {
        Self {
            state: Arc::new(Mutex::new(PlayerState::Idle)),
            actions: Vec::new(),
            current_index: 0,
            current_repeat: 0,
            speed: 1.0,
            repeat_count: 1,
            infinite_loop: false,
            timeout_seconds: 0.0,
            window_offset: None,
            window_title: String::new(),
            window_hwnd: 0,
            window_run_mode: "normal".into(),
            local_group_manager: None,
            stop_flag: Arc::new(AtomicBool::new(false)),
            pause_event: Arc::new((std::sync::Condvar::new(), std::sync::Mutex::new(true))),
            start_time: Instant::now(),
            thread_handle: None,
            callbacks: HashMap::new(),
        }
    }

    pub fn state(&self) -> PlayerState {
        *self.state.lock().unwrap()
    }

    pub fn set_actions(&mut self, actions: Vec<Action>) {
        if self.state() == PlayerState::Idle {
            self.actions = actions;
        }
    }

    pub fn set_speed(&mut self, speed: f64) {
        self.speed = speed.clamp(0.1, 10.0);
    }

    pub fn set_repeat_count(&mut self, count: i32) {
        self.repeat_count = count.max(1);
    }

    pub fn set_infinite_loop(&mut self, enabled: bool) {
        self.infinite_loop = enabled;
    }

    pub fn set_timeout(&mut self, seconds: f64) {
        self.timeout_seconds = seconds.max(0.0);
    }

    pub fn set_window_offset(&mut self, offset: Option<(i32, i32)>) {
        self.window_offset = offset;
    }

    pub fn set_window_title(&mut self, title: &str) {
        self.window_title = title.to_string();
    }

    pub fn set_window_hwnd(&mut self, hwnd: i64) {
        self.window_hwnd = hwnd;
    }

    pub fn set_window_run_mode(&mut self, mode: &str) {
        let allowed = ["normal", "offscreen", "hidden_taskbar", "offscreen_hidden_taskbar"];
        self.window_run_mode = if allowed.contains(&mode) { mode.to_string() } else { "normal".into() };
    }

    pub fn set_local_group_manager(&mut self, manager: LocalActionGroupManager) {
        self.local_group_manager = Some(Arc::new(Mutex::new(manager)));
    }

    pub fn play(&mut self) {
        if self.state() == PlayerState::Playing {
            return;
        }
        if self.actions.is_empty() {
            return;
        }

        if self.state() == PlayerState::Paused {
            let (cv, lock) = &*self.pause_event;
            let mut paused = lock.lock().unwrap();
            *paused = true;
            cv.notify_all();
            *self.state.lock().unwrap() = PlayerState::Playing;
            return;
        }

        self.stop_flag.store(false, Ordering::SeqCst);
        {
            let (cv, lock) = &*self.pause_event;
            let mut paused = lock.lock().unwrap();
            *paused = true;
            cv.notify_all();
        }
        *self.state.lock().unwrap() = PlayerState::Playing;
        self.start_time = Instant::now();

        let state = Arc::clone(&self.state);
        let stop_flag = Arc::clone(&self.stop_flag);
        let pause_event = Arc::clone(&self.pause_event);
        let actions = self.actions.clone();
        let speed = self.speed;
        let repeat_count = self.repeat_count;
        let infinite_loop = self.infinite_loop;
        let timeout_seconds = self.timeout_seconds;
        let start_time = self.start_time;

        self.thread_handle = Some(thread::spawn(move || {
            run_loop(state, stop_flag, pause_event, actions, speed, repeat_count, infinite_loop, timeout_seconds, start_time);
        }));
    }

    pub fn pause(&mut self) {
        if self.state() == PlayerState::Playing {
            let (cv, lock) = &*self.pause_event;
            let mut paused = lock.lock().unwrap();
            *paused = false;
            *self.state.lock().unwrap() = PlayerState::Paused;
        }
    }

    pub fn resume(&mut self) {
        if self.state() == PlayerState::Paused {
            self.play();
        }
    }

    pub fn toggle_pause(&mut self) -> PlayerState {
        match self.state() {
            PlayerState::Playing => {
                self.pause();
                self.state()
            }
            PlayerState::Paused => {
                self.resume();
                self.state()
            }
            other => other,
        }
    }

    pub fn stop(&mut self) {
        if matches!(self.state(), PlayerState::Playing | PlayerState::Paused) {
            self.stop_flag.store(true, Ordering::SeqCst);
            let (cv, lock) = &*self.pause_event;
            let paused = lock.lock().unwrap();
            cv.notify_all();
            *self.state.lock().unwrap() = PlayerState::Stopped;
        }
    }

    pub fn stop_and_wait(&mut self, timeout: Duration) -> bool {
        if matches!(self.state(), PlayerState::Playing | PlayerState::Paused) {
            self.stop_flag.store(true, Ordering::SeqCst);
            let (cv, lock) = &*self.pause_event;
            let _paused = lock.lock().unwrap();
            cv.notify_all();

            if let Some(handle) = self.thread_handle.take() {
                if handle.join().is_err() {
                    return false;
                }
            }
            *self.state.lock().unwrap() = PlayerState::Idle;
            true
        } else {
            true
        }
    }

    pub fn is_playing(&self) -> bool {
        self.state() == PlayerState::Playing
    }

    pub fn is_paused(&self) -> bool {
        self.state() == PlayerState::Paused
    }
}

fn run_loop(
    state: Arc<Mutex<PlayerState>>,
    stop_flag: Arc<AtomicBool>,
    pause_event: Arc<(std::sync::Condvar, std::sync::Mutex<bool>)>,
    actions: Vec<Action>,
    speed: f64,
    repeat_count: i32,
    infinite_loop: bool,
    timeout_seconds: f64,
    start_time: Instant,
) {
    let mut repeat = 0;

    loop {
        if stop_flag.load(Ordering::SeqCst) {
            break;
        }

        if timeout_seconds > 0.0 {
            if start_time.elapsed().as_secs_f64() >= timeout_seconds {
                break;
            }
        }

        if !infinite_loop && repeat >= repeat_count as usize {
            break;
        }

        for (i, action) in actions.iter().enumerate() {
            if stop_flag.load(Ordering::SeqCst) {
                break;
            }

            {
                let (cv, lock) = &*pause_event;
                let mut paused = lock.lock().unwrap();
                while !*paused {
                    if stop_flag.load(Ordering::SeqCst) {
                        break;
                    }
                    paused = cv.wait(paused).unwrap();
                }
                if stop_flag.load(Ordering::SeqCst) {
                    break;
                }
            }

            if timeout_seconds > 0.0 && start_time.elapsed().as_secs_f64() >= timeout_seconds {
                break;
            }

            if !action.check_condition(&HashMap::new()) {
                continue;
            }

            let delay_before = if speed > 0.0 { action.delay_before / speed } else { action.delay_before };
            if delay_before > 0.0 {
                interruptible_sleep(Duration::from_secs_f64(delay_before), &stop_flag);
            }

            if stop_flag.load(Ordering::SeqCst) {
                break;
            }

            let _ = execute_action(action);

            let delay_after = if speed > 0.0 { action.delay_after / speed } else { action.delay_after };
            if delay_after > 0.0 {
                interruptible_sleep(Duration::from_secs_f64(delay_after), &stop_flag);
            }
        }

        repeat += 1;
    }

    *state.lock().unwrap() = PlayerState::Idle;
}

fn interruptible_sleep(duration: Duration, stop_flag: &AtomicBool) {
    let end = Instant::now() + duration;
    while Instant::now() < end {
        if stop_flag.load(Ordering::SeqCst) {
            return;
        }
        let remaining = end.saturating_duration_since(Instant::now());
        let sleep_time = remaining.min(Duration::from_millis(50));
        if !sleep_time.is_zero() {
            thread::sleep(sleep_time);
        }
    }
}

fn execute_action(action: &Action) -> Result<(), String> {
    use simplerpa_winapi::input::InputSimulator;
    use simplerpa_winapi::background::BackgroundClicker;

    let simulator = InputSimulator::new();

    match action.action_type {
        ActionType::MouseClick => {
            let x = action.params.get("x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let y = action.params.get("y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let button = action.params.get("button").and_then(|v| v.as_str()).unwrap_or("left");

            if action.background_mode {
                let clicker = BackgroundClicker::new(Some(action.window_title.as_ref().map(|_| 0).unwrap_or(0)), action.window_title.as_deref());
                if clicker.is_available() {
                    let result = clicker.click(x, y, button, true);
                    if result.success { return Ok(()) }
                }
            }
            simulator.click(x, y, button);
            Ok(())
        }
        ActionType::MouseDoubleClick => {
            let x = action.params.get("x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let y = action.params.get("y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            simulator.double_click(x, y);
            Ok(())
        }
        ActionType::MouseRightClick => {
            let x = action.params.get("x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let y = action.params.get("y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            simulator.click(x, y, "right");
            Ok(())
        }
        ActionType::MouseMove => {
            let x = action.params.get("x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let y = action.params.get("y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            simulator.move_to(x, y);
            Ok(())
        }
        ActionType::MouseScroll => {
            let x = action.params.get("x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let y = action.params.get("y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let clicks = action.params.get("clicks").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            simulator.scroll(x, y, clicks);
            Ok(())
        }
        ActionType::KeyPress => {
            let key = action.params.get("key").and_then(|v| v.as_str()).unwrap_or("");
            simulator.key_press(key);
            Ok(())
        }
        ActionType::KeyType => {
            let text = action.params.get("text").and_then(|v| v.as_str()).unwrap_or("");
            let interval = action.params.get("interval").and_then(|v| v.as_f64()).unwrap_or(0.0);
            simulator.type_text(text, (interval * 1000.0) as u64);
            Ok(())
        }
        ActionType::Hotkey => {
            let keys: Vec<&str> = action.params.get("keys")
                .and_then(|v| v.as_array())
                .map(|arr| arr.iter().filter_map(|k| k.as_str()).collect())
                .unwrap_or_default();
            simulator.hotkey(&keys);
            Ok(())
        }
        ActionType::Wait => {
            let seconds = action.params.get("seconds").and_then(|v| v.as_f64()).unwrap_or(1.0);
            thread::sleep(Duration::from_secs_f64(seconds));
            Ok(())
        }
        ActionType::MouseMoveRelative | ActionType::MouseClickRelative => {
            let x = action.params.get("x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let y = action.params.get("y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;

            if action.background_mode {
                let clicker = BackgroundClicker::new(None, action.window_title.as_deref());
                if clicker.is_available() {
                    if action.action_type == ActionType::MouseMoveRelative {
                        let result = clicker.move_mouse(x, y, true);
                        if result.success { return Ok(()) }
                    } else {
                        let button = action.params.get("button").and_then(|v| v.as_str()).unwrap_or("left");
                        let result = clicker.click(x, y, button, true);
                        if result.success { return Ok(()) }
                    }
                }
            }
            Err("相对坐标操作需要窗口偏移量".into())
        }
        ActionType::ImageClick | ActionType::ImageWaitClick | ActionType::ImageCheck => {
            Err("图片识别需要 GUI 层的图像匹配支持".into())
        }
        ActionType::Screenshot => {
            Err("截图需要 GUI 层支持".into())
        }
        ActionType::ActionGroupRef => {
            let group_name = action.params.get("group_name").and_then(|v| v.as_str()).unwrap_or("");
            if group_name.is_empty() {
                return Err("未指定动作组名称".into());
            }
            // Action group execution needs to be handled by the GUI layer
            // as it requires access to the group manager and recursive execution
            Ok(())
        }
        ActionType::MouseDrag => {
            let sx = action.params.get("start_x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let sy = action.params.get("start_y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let ex = action.params.get("end_x").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            let ey = action.params.get("end_y").and_then(|v| v.as_i64()).unwrap_or(0) as i32;
            simulator.move_to(sx, sy);
            thread::sleep(Duration::from_millis(50));
            // Simple drag: move step by step with button held
            let steps = 10;
            for i in 1..=steps {
                let cx = sx + (ex - sx) * i / steps;
                let cy = sy + (ey - sy) * i / steps;
                simulator.move_to(cx, cy);
                thread::sleep(Duration::from_millis(20));
            }
            Ok(())
        }
    }
}
